### PR TITLE
Rockchip: add mainline linux support

### PR DESCRIPTION
**Pull request by Kwiboo**
This PR adds initial support for mainline Linux ~v5.1.9~ v5.4 for the Rockchip project.

HW decoding is supported for MPEG-2, VP8 and H264.
Initial YCbCr, 4K and 10-bit HDMI output is supported on RK3328.

~Only build tested so far, needs to be runtime tested.~

### CODE REVIEW

**Comment by MilhouseVH** [comment]
Tinkerboard and MiQi are missing `linux.arm.conf` for `default` kernel.

This reveals a small issue when sourcing the `linux` package, ie:
```
PROJECT=Rockchip DEVICE=MiQi ARCH=arm tools/pkginfo linux
```
as `kernel_config_path()` will return an empty string which then results in `grep` being [executed without a file](https://github.com/LibreELEC/LibreELEC.tv/blob/a5b603019bc310a0b17db7728ea754cb95259b08/packages/linux/package.mk#L47), which then hangs the package sourcing process.

We should have `kernel_config_path()` die() if there is no config available, and give the user a sensible error message. I'll PR a fix.

**Comment by MilhouseVH** [comment]
RK3399 `linux.aarch64.conf` is missing options and needs `make oldconfig` run on it, as it's prompting for options and halting the build.

RK3328 has built successfully.

**Comment by Kwiboo** [comment]
I will update RK3399 linux config, thanks for testing!

The `MiQi` and `TinkerBoard` is replaced by a `RK3288` device, mayby I should sunset the older devices in this PR.

**Comment by MilhouseVH** [comment]
Yes, if RK3288 is replacing those other devices then removing the old devices is probably the best option.

BTW this is the config diff I get when running `make oldconfig` for RK3399: http://ix.io/1LPD

**Comment by MilhouseVH** [comment]
RK3399 has built successfully (with the updated config) and generated all uboot images.

RK3328:
```
-rw-rw-r--  1 ubuntu ubuntu 117993479 Jun 15 08:02 LibreELEC-RK3328.arm-9.1-devel-20190615063153-#0615f-gf93f217-box.img.gz
-rw-rw-r--  1 ubuntu ubuntu       139 Jun 15 08:02 LibreELEC-RK3328.arm-9.1-devel-20190615063153-#0615f-gf93f217-box.img.gz.sha256
-rw-rw-r--  1 ubuntu ubuntu 117993308 Jun 15 08:03 LibreELEC-RK3328.arm-9.1-devel-20190615063153-#0615f-gf93f217-box-trn9.img.gz
-rw-rw-r--  1 ubuntu ubuntu       144 Jun 15 08:04 LibreELEC-RK3328.arm-9.1-devel-20190615063153-#0615f-gf93f217-box-trn9.img.gz.sha256
-rw-rw-r--  1 ubuntu ubuntu 117993448 Jun 15 08:03 LibreELEC-RK3328.arm-9.1-devel-20190615063153-#0615f-gf93f217-box-z28.img.gz
-rw-rw-r--  1 ubuntu ubuntu       143 Jun 15 08:03 LibreELEC-RK3328.arm-9.1-devel-20190615063153-#0615f-gf93f217-box-z28.img.gz.sha256
-rw-r--r--  1 ubuntu ubuntu  15591432 Jun 15 08:01 LibreELEC-RK3328.arm-9.1-devel-20190615063153-#0615f-gf93f217.kernel
-rw-rw-r--  1 ubuntu ubuntu  71406847 Jun 15 08:04 LibreELEC-RK3328.arm-9.1-devel-20190615063153-#0615f-gf93f217.log
-rw-rw-r--  1 ubuntu ubuntu 117993966 Jun 15 08:03 LibreELEC-RK3328.arm-9.1-devel-20190615063153-#0615f-gf93f217-roc-cc.img.gz
-rw-rw-r--  1 ubuntu ubuntu       142 Jun 15 08:03 LibreELEC-RK3328.arm-9.1-devel-20190615063153-#0615f-gf93f217-roc-cc.img.gz.sha256
-rw-rw-r--  1 ubuntu ubuntu 117993318 Jun 15 08:04 LibreELEC-RK3328.arm-9.1-devel-20190615063153-#0615f-gf93f217-rock64.img.gz
-rw-rw-r--  1 ubuntu ubuntu       142 Jun 15 08:04 LibreELEC-RK3328.arm-9.1-devel-20190615063153-#0615f-gf93f217-rock64.img.gz.sha256
-rw-rw-r--  1 ubuntu ubuntu 117993480 Jun 15 08:02 LibreELEC-RK3328.arm-9.1-devel-20190615063153-#0615f-gf93f217-rockbox.img.gz
-rw-rw-r--  1 ubuntu ubuntu       143 Jun 15 08:02 LibreELEC-RK3328.arm-9.1-devel-20190615063153-#0615f-gf93f217-rockbox.img.gz.sha256
-rw-r--r--  1 ubuntu ubuntu 108294144 Jun 15 08:01 LibreELEC-RK3328.arm-9.1-devel-20190615063153-#0615f-gf93f217.system
-rw-rw-r--  1 ubuntu ubuntu 125132800 Jun 15 08:04 LibreELEC-RK3328.arm-9.1-devel-20190615063153-#0615f-gf93f217.tar
-rw-rw-r--  1 ubuntu ubuntu       132 Jun 15 08:01 LibreELEC-RK3328.arm-9.1-devel-20190615063153-#0615f-gf93f217.tar.sha256
```
RK3399:
```
-rw-r--r--  1 ubuntu ubuntu  15591432 Jun 15 09:07 LibreELEC-RK3399.arm-9.1-devel-20190615081850-#0615g-g10395eb.kernel
-rw-rw-r--  1 ubuntu ubuntu 127016069 Jun 15 09:09 LibreELEC-RK3399.arm-9.1-devel-20190615081850-#0615g-g10395eb-khadas-edge.img.gz
-rw-rw-r--  1 ubuntu ubuntu       147 Jun 15 09:10 LibreELEC-RK3399.arm-9.1-devel-20190615081850-#0615g-g10395eb-khadas-edge.img.gz.sha256
-rw-rw-r--  1 ubuntu ubuntu  26850106 Jun 15 09:10 LibreELEC-RK3399.arm-9.1-devel-20190615081850-#0615g-g10395eb.log
-rw-rw-r--  1 ubuntu ubuntu 127015665 Jun 15 09:09 LibreELEC-RK3399.arm-9.1-devel-20190615081850-#0615g-g10395eb-rock960.img.gz
-rw-rw-r--  1 ubuntu ubuntu       143 Jun 15 09:09 LibreELEC-RK3399.arm-9.1-devel-20190615081850-#0615g-g10395eb-rock960.img.gz.sha256
-rw-rw-r--  1 ubuntu ubuntu 127015684 Jun 15 09:08 LibreELEC-RK3399.arm-9.1-devel-20190615081850-#0615g-g10395eb-rock-pi-4.img.gz
-rw-rw-r--  1 ubuntu ubuntu       145 Jun 15 09:08 LibreELEC-RK3399.arm-9.1-devel-20190615081850-#0615g-g10395eb-rock-pi-4.img.gz.sha256
-rw-rw-r--  1 ubuntu ubuntu 127015645 Jun 15 09:08 LibreELEC-RK3399.arm-9.1-devel-20190615081850-#0615g-g10395eb-rockpro64.img.gz
-rw-rw-r--  1 ubuntu ubuntu       145 Jun 15 09:08 LibreELEC-RK3399.arm-9.1-devel-20190615081850-#0615g-g10395eb-rockpro64.img.gz.sha256
-rw-rw-r--  1 ubuntu ubuntu 127015068 Jun 15 09:09 LibreELEC-RK3399.arm-9.1-devel-20190615081850-#0615g-g10395eb-sapphire.img.gz
-rw-rw-r--  1 ubuntu ubuntu       144 Jun 15 09:09 LibreELEC-RK3399.arm-9.1-devel-20190615081850-#0615g-g10395eb-sapphire.img.gz.sha256
-rw-r--r--  1 ubuntu ubuntu 117198848 Jun 15 09:07 LibreELEC-RK3399.arm-9.1-devel-20190615081850-#0615g-g10395eb.system
-rw-rw-r--  1 ubuntu ubuntu 134666240 Jun 15 09:10 LibreELEC-RK3399.arm-9.1-devel-20190615081850-#0615g-g10395eb.tar
-rw-rw-r--  1 ubuntu ubuntu       132 Jun 15 09:07 LibreELEC-RK3399.arm-9.1-devel-20190615081850-#0615g-g10395eb.tar.sha256
```

**Comment by MilhouseVH** [comment]
RK3328 also needs these config options (for some reason - unlike RK3399 - it didn't halt the build): http://ix.io/1LPS

**Comment by mo123** [comment]
Please check gigabit ethernet on RK3328 devices like ROC CC, TRN9, ethernet is not working for these devices on mainline images.
I see the CPU speed for RK3328 is also limited to 1.2Ghz in mainline and in Rockchip 4.4 kernels, perhaps 1.4/1.5Ghz support can be added back if it runs stable to help with 4K and software decoding?
If possible also change the default CPU governor to interactive instead of ondemand, it runs better and doesn't cause CPU spikes if going from low to high frequencies.

**Comment by Kwiboo** [comment]
@mo123 I am working on an update of this PR and addressing ethernet issue on roc-cc and trn9 is on the list of things I want to check/fix.

Regarding higher cpu frequency then 1.3 GHz on rk3328, that is not something I am going to push for.
Even if datasheet lists 1.4 GHz as max, Rockchip only recommend using 1.3 GHz as the maximum for stable operation, see https://patchwork.kernel.org/patch/10835789/#22513895.
Ideally dtoverlays can be used in future to allow overclocking.

The interactive gov is only part of android kernels and not part of mainline linux. #3538 could possible help tune the on demand gov.

**Comment by Kwiboo** [comment]
PR has been updated to linux v5.3, it includes RK and V4L2 backports from v5.4 and linux-next.
HW decoding for MPEG-2, VP8 and H264 is supported.

Needs some more testing on devices beside Rock64 and Tinker Board S before I can mark this ready for review.

File: projects/Rockchip/devices/RK3288/options [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  # Additional kernel make parameters (for example to specify the u-boot loadaddress)
+    KERNEL_MAKE_EXTRACMD=""
+    KERNEL_MAKE_EXTRACMD+=" rk3288-miqi.dtb"
+
+    if [ "$LINUX" = "rockchip-4.4" ]; then
+      KERNEL_MAKE_EXTRACMD+=" rk3288-miniarm.dtb"
+    else
+      KERNEL_MAKE_EXTRACMD+=" rk3288-tinker.dtb"
+      KERNEL_MAKE_EXTRACMD+=" rk3288-tinker-s.dtb"
+    fi
```
[the remark this answers is not in the document]
Comment (Kwiboo) [reply projects/Rockchip/devices/RK3288/options]: I guess they are, there was too many dtbs last time I tested building all, Tinker Board is the only RK3288 device I can test on. I will test building with all dtbs once my current clean build completes.

**Comment by butterkitty** [comment]
I've been testing using my RockPI4 4gb with 32GB SDCard with HDMI connection to audio receiver/TV.  I wasn't sure where to put the writeup, so just let me know :)

Tested working:
  Video HW Decoding
    MPEG2
    H264
  Audio Passthrough:
    AC3
    DTS

Tested **not** working:
  Video HW Decoding
    H265 (not expected)
    VP9 (not expected)
  Audio Passthrough:
    DTS-HD
    Dolby Live

Other Problems
  Wifi on boot doesn't work sometimes.  Requires a reset of the wifi via the LibreElec app inside of Kodi.  (happens on 2.4ghz and 5ghz)
  Audio pops (not unlike a record player) when not in passthrough mode over HDMI and when in passthrough mode audio cuts very very slightly randomly.

**Comment by KOPRajs** [comment]
@butterkitty Can you please look at this bug report and confirm whether the HDMI audio issues are the same: https://forum.libreelec.tv/thread/17600-rk3399-hdmi-audio-quality/

**Comment by butterkitty** [comment]
@KOPRajs That is exactly the issue I seem to be having and I've commented on that thread accordingly. 

Thank you!

**Comment by Kwiboo** [comment]
This PR has once again been updated, this time to linux v5.4, it includes RK and V4L2 backports from v5.5 and linux-next.
HW decoding for MPEG-2, VP8 and H264 is supported.

I have only completed runtime test on Tinker Board S, will continue on my other boards and mark this ready for review once I know all my devices boot and start kodi.

**Comment by zeus24** [comment]
Hi, can you check the build parameters for the kernel?
You disabled PCI und NVM-E support, but the RockChip Pro has a PCI Express ports and many people are using this for a NVM-E SSD.

**Comment by Kwiboo** [comment]
I have pushed a final update for 2019, let’s bring Rockchip mainline into the new decade and leave the old RK 4.4 in the past!

I have boot tested the latest update on the following devices:
- RK3288: Tinker Board S
- RK3328: Rock64, ROC-CC, Rockbox, MVR9
- RK3399: Rock Pi 4, Sapphire

Early test images can be found at http://kwiboo.libreelec.tv/test/

There is experimental patches included for initial YCbCr, 4K and 10-bit HDMI output support on RK3328. The color space conversion parts have not been adopted for YCbCr and 10-bit output so wrong color encoding and range is expected, just one of many things to improve in the future.

@zeus24 Thanks for testing! I have left PCI and NVM-E support disabled for now as I cannot test anything PCIe related at this moment and I really want this PR merged sooner rather than later.

**Pull request merged**